### DISCUSSION TITLE
Idea how to ligthen 6581 filter woes

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [libsidplayfp/libsidplayfp#93] [Ideas]
Unfortunately, there is no way to get the 6581 filter "right", as there is too much variation even between chips made in the same factory on the same day.

My idea centers around the logic that most composers from the golden age used their C64 their entire career. 

So why not create a small lookup table for each composer (at least for the big ones) that adjusts the filter settings for the 6581?

e.g. for David Dunn I would personally set the filterCurve to 1.0. Simply load the tune "Jump Jet", and you will hear that voices 2 & 3 are almost silent from 0:20 to 0:38 and then again from 0:45 to 0:51. If you set the filter curve to 1.0, suddenly, you can clearly hear the voices playing. If you listen to the other tunes from David Dunn, you will find that all of them sound much better.

For Martin Galway it's a similar story. There, I found a setting of 0.6 - 0.65 to be the most pleasant. You can hear the filter doing something, whereas it sounds a bit thin/bland with the defaults.

Nobody can claim they've ever heard the intended versions unless they composed it, so I propose tuning the filter settings by ear for the biggest 20-30 composers that used 6581. They should be tweaked so that the filter FX are not too subtle to be almost (or entirely) inaudible but also not so strong that the volume between voices becomes unbalanced.

I would be willing to write a small class that returns "recommended" filterCurve settings when feeding in a composer name. It would just be an ```std::unordered_map``` that returns 0.5 (or -1.0?) for unrecognized names.

I would also love to provide these tunings, listen to feedback, etc., but I want to ensure that I didn't forget anything and that my idea makes sense.

I think this would be a very welcome "fix" that, for the first time, actually represents the artist's intent. Everyone would hear these old tunes fresh for the first time, the way they were (probably) intended.

Somebody will undoubtedly mention that we could contact the original authors and ask for their input, but I have my doubts about that. Some are no longer reachable (deceased, moved on, and don't care anymore, no contact info available, etc.), and I don't think we can rely on their memory anyway, as these tunes were made 30-35 years ago.

**Comment by mrdudz** [comment]
Didn't Alankila collect the filter characteristics of the C64s from a bunch of people ages ago, and didn't resid-fp already have presets for those? At least we had them in VICE when we still used resid-FP

**Reply by Toddler-Boy** [reply]
> At least we had them in VICE when we still used resid-FP

Can you tell me why you switched and to what? I'm very curious.

**Reply by mrdudz** [reply]
Basically,. the regular resid got a huge update from Dag Lem (the original Author), and then Alankila removed resid-FP, because there was little reason to keep it around.

**Reply by Toddler-Boy** [reply]
Was that recently? Did it improve accuracy? It seems a lot of extra work went into libsidplayfp to make the filters behave more like the real thing.

**Reply by mrdudz** [reply]
That was many years ago. Generally the original resid is more "accurate" in how it works internally, resid-FP took a different approach and "tweaked" the filters instead. Today there shouldn't be a big audible difference between the two. Much less of a difference than between individual SIDs anyway :)

**Reply by Toddler-Boy** [reply]
What does the internal frame-buffer format for VICE look like? One byte per pixel (0-15)? I've written a small CRT emulation using OpenGL shaders that looks a LOT nicer and like the real thing. Adjustable curvature, chroma subsampling, brightness/contrast/saturation, bloom, flicker, noise, etc., and even chroma into luma bleeding. I would be very interested in helping make the CRT emulation output of VICE look nicer.

It also correctly handles the differences between PAL and NTSC (Hanover bars and their suppression via a delay line). All are implemented with fragment shaders, and the GPU usage is extremely low. Only around 3ms on my GTX 1060.

**Reply by mrdudz** [reply]
This is very offtopic now :)

One byte per pixel, yes. It'd take a bit of effort to make this happen though (it should work for all emulators, for all OSs, for all ports - that may mean making 2 or 3 different implementations of that filter) . I want to at least add one more iteration of the software filter before that happens (one that renders in triple size with a dot matrix texture) though.

you may join #vice-dev on libera.chat (late evenings EST)

**Comment by drfiemost** [comment]
> Didn't Alankila collect the filter characteristics of the C64s from a bunch of people ages ago, and didn't resid-fp already have presets for those? At least we had them in VICE when we still used resid-FP

That was an old version, the filter setting has been later simplified to a single value.

The idea is interesting, although I think it belongs to the frontend and not to the library. Also the settings would be subjective and depend on personal preferences, but not a big issue if it remains an optional feature.
It can be implemented as an _AutoFilter_ option in sidplayfp, which will set the filter parameter based on the tune author, and maybe extend it to tune specific setting in a future revision if needed.

**Comment by Toddler-Boy** [comment]
> The idea is interesting, although I think it belongs to the frontend and not to the library.

Since it's such a small thing (only like 10 lines of code and a small table), I would put it into the SidTuneInfo class. But that's because my fork is essentially a "libsidplayEZ", so easy to use. The goal is you instantiate one class (the player), and it will have functions to load tunes and use the correct settings (NTSC/PAL, 6581/8580, VBI/CIA, etc.) automatically.

**Comment by Toddler-Boy** [comment]
I've played around with this, and a pattern is emerging. The current filter settings for the 6581 are a bit too conservative. If I change the default from 0.5 to 0.6, the majority of tested authors' tunes immediately sound a lot better. The filter movement is better audible, making the whole tune more animated and more enjoyable to listen to. I'm currently re-re-discovering (not a typo) SID music 🥲

Also, the filter distorts a bit too fast and too much. That might be a side effect of the filter curve being adjusted so much. For all of "David Dunn"s tunes, the curve sounds best at 1.0, but the distortion when more than one voice is using the filter is very distracting and clearly not intended.

Maybe the filter can be changed so that instead of just having a single parameter, it can have multiple parameters (cutoff offset, resonance offset, distortion threshold, distortion depth, etc.) to fine-tune this process more.

**Reply by drfiemost** [reply]
The filter is not perfect, a few parameters have been tweaked empirically (the voltages and uCox) and the opamp curve have been measured on a 6581R4AR  and might not be representative of all revisions. Also, going technical, the EKV transistor model is missing the pinch voltage. I tried to implement it but the results were disappointing so it's disabled.
The parameter only affects the filter cutoff DAC. It would be possible to make some values configurable but then you need to know a realistic range and some combinations could also cause a crash.

**Comment by drfiemost** [comment]
According to Antti Lankila's analysis https://github.com/libsidplayfp/libsidplayfp/issues/66#issuecomment-1456137641 Galway's filter, and even more Follin's, are lighter than average, so a value less than 0.5 should be used.

**Reply by Toddler-Boy** [reply]
It seems than that the filter-curve parameter is applied in reverse. The larger the curve value, the smaller the cutoff becomes. So 0.6 is actually 0.4, 0.7 is 0.3, etc.

**Reply by drfiemost** [reply]
Ah right, in residfp the logic is inverted compared to resid, as the [documentation](https://github.com/libsidplayfp/sidplayfp/wiki/sidplayfp.ini#emulation) says .

**Reply by Toddler-Boy** [reply]
Since residfp is the odd one out here, I would like that logic inverted so it feels correct. Imagine someone uses certain players, and the logic suddenly reverted only for this one. Data exchange (like our filter suggestion table) becomes harder. Plus, it's really confusing to me that I increase a value and the cutoff moves down.

**Reply by drfiemost** [reply]
I don't know how this happened in the first place, but unless there are some bigger filter changes it will stay as is. Changing things at this point would be even more confusing.
